File: my-task-manager-java/src/manager.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class manager {
    static Scanner scan = new Scanner(System.in);
    static ArrayList<String> tasks =
            new ArrayList<>(Arrays.asList("task1", "task2", "task3"));

    static void displayTasks(ArrayList<String> array) {
        System.out.println("---");
        for(String element : array) {
            System.out.println(element);
        }
    }

    static void continueOrExit(String option) {
        switch (option) {
            case "0":
                exit();
                break;
            default:
                showMain();
                break;
        }
    }

    static boolean isEmpty(ArrayList<String> array) {
        return array.size() == 0;
    }

    static void noTasks() {
        System.out.println("You have no tasks yet! " +
                "Continue [any key] - Exit [0]");
        String option = scan.nextLine();
        continueOrExit(option);
    }

    static void showTasks() {
        if(isEmpty(tasks)) {
            noTasks();
        } else {
            System.out.println("Here are your tasks " +
                    "Continue [any key] - Exit[0]:\n");
            displayTasks(tasks);
            String option = scan.nextLine();
            continueOrExit(option);
        }
    }

    static void askForNewAddTaskChoice (String option) {
        switch (option) {
            case "1":
                addTask();
                break;
            default:
                showMain();
                break;
        }
    }

    static void addTask() {
        displayTasks(tasks);
        System.out.println("Which task do you want to add? " +
                "(Cancel [0])");
        String task = scan.nextLine();
        if(task == "0") {
            showMain();
        } else if (task.isEmpty()) {
            addTask();
        } else {
            tasks.add(" " + task);
            displayTasks(tasks);
            System.out.println("Add another task? " +
                    "(Yes [1] - No [any key])");
            String option = scan.nextLine();
            askForNewAddTaskChoice(option);
        }
    }

    /*static void askForNewDeleteTaskChoice (String option) {
        switch (option) {
            case "1":
                deleteTask();
                break;
            default:
                showMain();
                break;
        }
    }*/

    static boolean outOfArray(ArrayList<String> array, int index) {
        return index < 0 || index > array.size();
    }

    static void deleteTask() {
        if(isEmpty(tasks)) {
            noTasks();
        } else {
            displayTasks(tasks);
            System.out.println("Which task do you want to delete? " +
                    "(give an index begginning from 1 - Cancel [0])");
            int option = scan.nextInt();
            if(option == 0) {
                showMain();
            } else if(outOfArray(tasks, option)) {
                deleteTask();
            } else {
                tasks.remove(option - 1);
                if(isEmpty(tasks)) {
                    showMain();
                } else {
                    /*System.out.println("Delete another task? " +
                            "(yes [1] - no [any key])");
                    String optionString = scan.nextLine();
                    askForNewDeleteTaskChoice(optionString);*/
                    deleteTask();
                }
            }
        }
    }

    /*static void askForNewMarkTaskDoneChoice (String option) {
        switch (option) {
            case "1":
                markTaskDone();
                break;
            default:
                showMain();
                break;
        }
    }*/

    static void markTaskDone() {
        displayTasks(tasks);
        System.out.println("Which task do you want to mark as done? " +
                "(give an index begginning from 1 - Cancel [0])");
        int option = scan.nextInt();
        if(option == 0) {
            showMain();
        } else if(outOfArray(tasks, option)) {
            markTaskDone();
        }else if(tasks.get(option - 1).contains("done")) {
            markTaskDone();
        } else {
            tasks.set(option - 1,
                    tasks.get(option - 1).concat(" (done)"));
            displayTasks(tasks);
            /*System.out.println("Which task do you want to mark as done? " +
                    "(give an index begginning from 1 - Cancel [0])");
            String optionString = scan.nextLine();
            askForNewMarkTaskDoneChoice(optionString);*/
            markTaskDone();
        }
    }

    static void exit() {
        System.out.println("Bye Bye ;-)");
        System.exit(0);
    }

    static void showMain() {
        System.out.println("Welcome to your task manager, Press:\n" +
                "1. to see all your tasks\n" +
                "2. to add a task\n" +
                "3. to delete a task\n" +
                "4. to mark a task as done\n" +
                "5. to Exit the task manager\n");
        String option = scan.nextLine();
        switch (option) {
            case "1":
                showTasks();
                break;
            case "2":
                addTask();
                break;
            case "3":
                deleteTask();
                break;
            case "4":
                markTaskDone();
                break;
            case "5":
                exit();
                break;
            default:
                showMain();
        }
    }

    public static void main(String[] args) {
        showMain();
    }
}

File: JavaScript/manager.js
const readline = require('readline');
var rl = readline.createInterface(process.stdin, process.stdout);

const fs = require('fs');
var data = fs.readFileSync('./JavaScript/tasks.json');
var tasks = JSON.parse(data);

const saveData = (newData) => {
    newData = JSON.stringify(newData);
    fs.writeFile('./JavaScript/tasks.json', newData, 'utf-8', (error) => {
        if(error)
            console.log(error);
    });
};

const displayTasks = (array) => {
    console.log("---")
    for(element of array) {
        console.log(element);
    }
}

const continueOrExit = (option) => {
    switch (option) {
        case "0":
            exit();
            break;
        default:
            showMain();
            break;
    };
};

const isEmpty = (array) => {
    return array.length == 0;
};

const noTasks = () => {
    rl.question("You have no tasks yet! (Continue [any key] - Exit [0])\n", (option) => {
        continueOrExit(option);
    });
};

const showTasks = () => {
    if(isEmpty(tasks.tasks)) {
        noTasks();
    } else {
        displayTasks(tasks.tasks);
        rl.question(`Here are your tasks (Continue [any key] - Exit [0]):\n`, (option) => {
            continueOrExit(option);
        });    
    }
};

const askForNewAddTaskChoice = (option) => {
    switch (option) {
        case "1":
            addTask();
            break;
        default:
            showMain();
            break;
    }
};

const addTask = () => {
    displayTasks(tasks.tasks);
    rl.question(`Which task do you want to add? (Cancel [0])\n`, (task) => {
        if(task === "0") {
            showMain();
        } else if (task === "") {
            addTask();
        } else {
            tasks.tasks.push(`${task}`);
            saveData(tasks);
            displayTasks(tasks.tasks);
            rl.question(`Add another task? (Yes [1] - No [any key])\n`, (option) => {
                askForNewAddTaskChoice(option);
            });
        }
    });
};

const askForNewDeleteTaskChoice = (option) => {
    switch (option) {
        case "1":
            deleteTask();
            break;
        default:
            showMain();
            break;
    }
};

const outOfArray = (array, index) => {
    return index < 0 || index > array.length;
};

const deleteTask = () => {
    if(isEmpty(tasks.tasks)) {
        noTasks();
    } else {
        displayTasks(tasks.tasks);
        rl.question(`Which task do you want to delete? (give an index begginning from 1 - Cancel [0])\n`, (option) => {
            option = parseInt(option);
            if(option == 0) {
                showMain();
            } else if(outOfArray(tasks.tasks, option)) {
                deleteTask();
            } else {
                tasks.tasks.splice(option - 1, 1);
                saveData(tasks);
                if(isEmpty(tasks.tasks)) {
                    showMain();
                } else {
                    displayTasks(tasks.tasks);
                    rl.question(`Delete another task? (yes [1] - no [any key])\n`, (option) => {
                        askForNewDeleteTaskChoice(option);
                    });        
                }
            }
        });    
    }
};

const askForNewMarkTaskDoneChoice = (option) => {
    switch (option) {
        case "1":
            markTaskDone();
            break;
        default:
            showMain();
            break;
    }
};

const markTaskDone = () => {
    displayTasks(tasks.tasks);
    rl.question(`Which task do you want to mark as done? (give an index begginning from 1 - Cancel [0])\n`, (option) => {
        option = parseInt(option);
        if(option == 0) {
            showMain();
        } else if(outOfArray(tasks.tasks, option)) {
            markTaskDone();
        } else if(tasks.tasks[option - 1].includes("done")) {
            markTaskDone();
        } else {
            tasks.tasks[option - 1] = tasks.tasks[option - 1].concat(" (done)");
            saveData(tasks);
            displayTasks(tasks.tasks);
            rl.question(`Mark another task as done? (yes [1] - no [any key])\n`, (option) => {
                askForNewMarkTaskDoneChoice(option);
            });        
        }
    });
};

const exit = () => {
    console.log("Bye bye ;-)");
    rl.close();
};

const showMain = () => {
    rl.question("Welcome to your task manager, Press:\n 1. to see all your tasks\n 2. to add a task\n 3. to delete a task\n 4. to mark a task as done\n 5. to Exit the task manager\n", (option) => {
        switch (option) {
            case "1":
                showTasks();
                break;
            case "2":
                addTask();
                break;
            case "3":
                deleteTask();
                break;
            case "4":
                markTaskDone();
                break;
            case "5":
                exit();
                break;
            default:
                showMain();
        }
    });
};

showMain();